### PR TITLE
Template inheritance

### PR DESCRIPTION
**Pull request by knoxsp**
More changes required to make the template inheritance more stable, namely not deleting all the data....

### CODE REVIEW

File: tests/templates/test_template_inheritance.py [hunk 1]
```diff
@@ -187,6 +187,111 @@ def test_add_typeattr_to_child(self, client):
...
+
+        #now add *the same* typeattr to the parent type. This should result in the child type returing its own typeattr
+        #and the parent type returning a different typeattr with the same attr_id
+        new_parent_typeattr = JSONObject({
+            'attr_id': newattr.id,
+            'type_id': type_to_update.id
+        })
+        client.add_typeattr(new_parent_typeattr)
+
+        #Fetch both the parent and child to verify that the child nd parent types
```
Comment (pmslavin) [inline tests/templates/test_template_inheritance.py:260]:
```suggestion
        #Fetch both the parent and child to verify that the child and parent types
```

File: tests/templates/test_template_inheritance.py [hunk 2]
```diff
@@ -187,6 +187,111 @@ def test_add_typeattr_to_child(self, client):
...
+            if templatetype.id == type_to_update.id:
+                un_updated_parent_type_attrs = templatetype.typeattrs
+                break
+
+        assert len(updated_child_type_attrs) == len(original_type_attrs)+1
+        assert len(updated_child_type_attrs) == len(un_updated_parent_type_attrs)+1
+
+
+        #now add *the same* typeattr to the parent type. This should result in the child type returing its own typeattr
+        #and the parent type returning a different typeattr with the same attr_id
```
Comment (pmslavin) [inline tests/templates/test_template_inheritance.py:253]:
```suggestion
        #and the parent type returning a different typeattr, i.e. a different type_id with the same attr_id
```

File: tests/templates/test_template_inheritance.py [hunk 3]
```diff
@@ -401,5 +506,57 @@ def test_create_network_with_child_template(self, client):
...
 
         new_network = client.add_network(network)
+        #Now do a get network to verify the network's there
+        requested_network = client.get_network(new_network.id)
+
+        assert requested_network.nodes[0].types[0].child_template_id == child_template_j.id
+
+        #Make sure that the correct attributes have been added from the child (inherihted
+        #from the parent)
+        assert len(requested_network.attributes) == len(parent_network_type.typeattrs)
```
Comment (pmslavin) [inline tests/templates/test_template_inheritance.py:516]: This checks that the _number_ of attributes is consistent.  Set equality would check that they are the same attrs,

File: tests/templates/test_template_inheritance.py [hunk 4]
```diff
@@ -401,5 +506,57 @@ def test_create_network_with_child_template(self, client):
...
+
+
+    def test_delete_parent_type(self, client):
+        """
+            Test to ensure that when you delete a parent type, its child types
+            are also deleted (or not depoending on the 'force' flag)
```
Comment (pmslavin) [inline tests/templates/test_template_inheritance.py:522]: Does the "force flag" in `delete_templatetype` referred to here mean that function's `force` or `delete_children` kwarg?  L541-542 below uses both terms.

Also, see the suggestion to `__init__.py` above regarding `force` vs `delete_resourcetypes`.

File: tests/templates/test_template_inheritance.py [hunk 5]
```diff
@@ -401,5 +506,57 @@ def test_create_network_with_child_template(self, client):
...
+
+
+        updated_child_template_j = client.get_template(child_template_j.id)
+        assert len(updated_child_template_j.templatetypes) == len(child_template_j.templatetypes) - 1
+        assert child_type.id not in [t.id for t in updated_child_template_j.templatetypes]
+        #all types in the child should now belong to the parent. The chid one is gone.
+        for tt in updated_child_template_j.templatetypes:
+            #there should always be a network type for the child
+            if tt.resource_type == 'NETWORK':
+                continue
```
Comment (pmslavin) [inline tests/templates/test_template_inheritance.py:557]: i.e.  The child should always have a distinct network type of its own, so the test on L560 below doesn't apply in that case.

File: tests/test_login.py [hunk 6]
```diff
@@ -85,10 +85,10 @@ def test_login_too_many_attempts(self, client):
...
                 user_id, session_id = client.login('root', 'wrong-password!')
-            except hydra_base.exceptions.HydraLoginUserMaxAttemptsExceeded as e:
+            except hydra_base.exceptions.HydraError as e:
                 # this exception will be eventually raised at a certain point
                 exception_raised="HydraLoginUserMaxAttemptsExceeded"
-            except hydra_base.exceptions.HydraLoginUserPasswordWrong as e:
+            except hydra_base.exceptions.HydraError as e:
                 pass
 
         assert exception_raised == "HydraLoginUserMaxAttemptsExceeded"
```
Comment (pmslavin) [inline tests/test_login.py:94]: The `HydraLoginUserMaxAttemptsExceeded` is no longer raised, so...
```python
                # this exception will be eventually raised at a certain point
                exception_raised="HydraLoginUserMaxAttemptsExceeded"
```
...isn't true.  Not sure what the "certain point" here is, or why there is the `exception_raised` var.

These could use `with pytest.raises(...)` in the same way as the code above.

**Review by pmslavin — COMMENTED** [review]
Few suggestions and questions.

File: tests/templates/test_template_inheritance.py (hunk 5 above)
Comment (knoxsp) [reply tests/templates/test_template_inheritance.py:557]: Correct

**Pull request merged**